feat(app): Add hostname label to http metrics

cratelyn opened the pull request:
this commit introduces an additional label to HTTP and gRPC route
metrics, containing the DNS host of requests. requests destined for an
ip address of some kind are not recorded with a hostname metric, as a
way to help minimize the impact of time series cardinality.

the core of this change is in this field addition:

```diff
// /linkerd/app/outbound/src/http/logical/policy/route/metrics/labels.rs
-pub struct Route(pub ParentRef, pub RouteRef);
+pub struct Route {
+    parent: ParentRef,
+    route: RouteRef,
+    hostname: Option<dns::Name>,
+}
```

see this part of the change to our `MkStreamLabel` implementation, used
in our metrics tracking request durtion, and counting response status
codes:

```diff
-    fn mk_stream_labeler<B>(&self, _: &::http::Request<B>) -> Option<Self::StreamLabel> {
+    fn mk_stream_labeler<B>(&self, req: &::http::Request<B>) -> Option<Self::StreamLabel> {
         let parent = self.params.parent_ref.clone();
         let route = self.params.route_ref.clone();
-        Some(metrics::LabelHttpRsp::from(metrics::labels::Route::from((
-            parent, route,
-        ))))
+        Some(metrics::LabelHttpRsp::from(metrics::labels::Route::new(
+            parent,
+            route,
+            req.uri(),
+        )))
     }
```

we now inspect the request, and use the URI to label metrics related to
this traffic by hostname.

a `http_request_hostnames()` test case is added to exercise this.

some todo comments are left, noting where we would ideally like to
simplify or generalize the machinery related to `RetryLabelExtract`, the
type that bridges the labels needed based on our `NewService<T>` target,
and the request type later accepted by the instantiated `Service<T>`.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route.rs:
```diff
@@ -174,15 +186,15 @@ impl<T> filters::Apply for Http<T> {
 impl<T> metrics::MkStreamLabel for Http<T> {
     type StatusLabels = metrics::labels::HttpRouteRsp;
-    type DurationLabels = metrics::labels::Route;
+    type DurationLabels = metrics::labels::HttpRoute; // TODO(kate): do we want hostname here?
```
right now we'd be adding the hostname label to our durations. if cardinality concerns are going to matter anywhere, they'll be happening here.

on the other hand though, i can imagine this being tremendously useful latency info.

i'd love to hear what people think!

cratelyn replied:
i've refrained from adding that label here.

it's still an interesting follow-on / enhancement to consider, but i'll go ahead and mark this as resolved for now.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route.rs:
```diff
@@ -174,15 +186,15 @@ impl<T> filters::Apply for Http<T> {
+    fn mk_stream_labeler<B>(&self, req: &::http::Request<B>) -> Option<Self::StreamLabel> {
         let parent = self.params.parent_ref.clone();
         let route = self.params.route_ref.clone();
-        Some(metrics::LabelHttpRsp::from(metrics::labels::Route::from((
-            parent, route,
-        ))))
+        Some(metrics::LabelHttpRsp::from(
+            metrics::labels::HttpRoute::from((parent, route, req.uri().host().map(str::to_owned))),
+        ))
     }
```
this is the part where the actual label value gets discovered and passed into our `LabelHttpRouteRsp` label set.

cratelyn reviewed (commented):
:speech_balloon: some notes...

cratelyn commented:
on monday i'm going to pick this up and update the tests, which is why ci is currently failing.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route.rs:
```diff
@@ -149,15 +149,27 @@ impl<T: Clone, M, F, P> svc::Param<BackendDistribution<T, F>> for MatchedRoute<T
+        HttpRouteLabels(
+            self.params.parent_ref.clone(),
+            self.params.route_ref.clone(),
+            // we do not know the hostname at instantiation time for the stack. this is found
+            // later when we are creating our labeler.
+            None,
```
on a further read, it seems like we want to use `svc::ExtractParam`

cratelyn replied:
another way, for the record, could be to store the hostname in our route match structure, but i considered that to be a little invasive.

cratelyn replied:
resolved in subsequent commits. :white_check_mark:

cratelyn commented:
checking in an incremental snapshot, it is not expected to build right now.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route/metrics/labels.rs:
```diff
@@ -50,15 +56,55 @@ pub enum Error {
+// === impl HttpRoute ===
+
+impl HttpRoute {
+    pub fn new(parent: ParentRef, route: RouteRef, host: Option<&str>) -> Self {
+        let host = match host.and_then(|h| url::Host::parse(h).ok()) {
+            Some(url::Host::Domain(s)) => Some(s),
+            // Refrain from emitting label values for url's whose host component are a IPv4
+            // or IPv6 address, to help prevent label cardinality from becoming too noisy.
+            None | Some(url::Host::Ipv4(_) | url::Host::Ipv6(_)) => None,
+        };
```
:point_up: this is where we filter out addresses from our labels

olix0r replied:
We already have some prior art in the proxy that avoids the url dependency and is oriented the hyper URI types. Can [`Addr`](https://github.com/linkerd/linkerd2-proxy/blob/89ba5b5d8b39f54de8a8c6e1e8cf552e9c9a5527/linkerd/addr/src/lib.rs#L50) work here?

cratelyn replied:
i took a look at the `Addr` types, but these seem subtly different from what we want here. those types are focused on working with the authority, which includes the port in addition to the hostname.

for what it is worth, we do have `url` as a transitive dependency in the workspace today, via the hickory-resolve crate, see: https://github.com/linkerd/linkerd2-proxy/pull/3258/commits/9e3ee3e6611ea96e18d74cf687a33a4c86eb04a8

my chief concern is that using the authority might exacerbate the cardinality problem, as we would have a time series counting e.g. each status code, for each port, for each host. that extra dimension could be costly, until we have some way to limit the number of time series we are emitting.

on the upside, this might help disambiguate between services running on shared infrastructure.

i'm not tremendously opinionated about hostname vs. authority though, and perhaps reusing the `Addr` crate is a nice benefit.

what do you think, @olix0r?

olix0r replied:
Ah, I see what you mean.

I don't think that cardinality is particularly a concern, since parent refs may generally include port information; and disambiguating services by port is pretty natural.

I think the bigger concern is how this compares with values in the TLS case, where we'll have ServerNames that omit ports.

But we still could avoid the url types with, e.g. [`NameAddr::name` ](https://github.com/linkerd/linkerd2-proxy/blob/89ba5b5d8b39f54de8a8c6e1e8cf552e9c9a5527/linkerd/addr/src/lib.rs#L222).

I.e. from an Addr you can get at it via something like `addr.name_addr().and_then(NameAddr::name)`.

This also ensures some amount of validation (i.e. over &str) where we can accept invalid hostnames.

cratelyn replied:
i've pushed aed9d303. this commit does two things: (1) avoids a `url` dependency, via use of `linkerd_app_core::dns::Name`, and (2) tweaks the interfaces of `http::logical::policy::route::metrics::labels::HttpRoute` to accept `http` types rather than an `Option<&str>`.

e2447dc0 then reverts our addition of a direct `url` dependency.

this bore out really nicely, thank you for taking the time to work through this! how does this look to you now, @olix0r?

cratelyn replied:
marking this as resolved, based on in-air conversation. :slightly_smiling_face:

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route/metrics/labels.rs:
```diff
@@ -5,17 +5,23 @@ use prometheus_client::encoding::*;
 
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
-pub struct Route(pub ParentRef, pub RouteRef);
+pub struct HttpRoute {
+    parent: ParentRef,
+    route: RouteRef,
+    host: Option<String>,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct GrpcRoute(pub ParentRef, pub RouteRef);
```
:point_up: this is where we add a hostname to HTTP metrics, but intentionally refrain from doing so for our gRPC metrics.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route.rs:
```diff
@@ -117,20 +116,37 @@ where
                 // AND/OR headers
                 .push(extensions::NewSetExtensions::layer())
                 .push(metrics::layer(&metrics.requests))
+                .check_new::<Self>()
+                .check_new_service::<Self, http::Request<linkerd_app_core::proxy::http::BoxBody>>()
```
:point_up: this is mentioned in the pr description. i am happy to walk these additions back; they made my own work easier, and it seemed like an easy gift to offer future engineers tweaking our policy stack. :gift:

sfleen replied:
I'm more than fine with these, the only real concern would be compile time impact but I don't expect these to be a burden here.

olix0r replied:
We do tend to prune tehse down to avoid noise, but it's not really a problem to leave them in. If we do leave them in, though, I'd strongly prefer that we rely on the reexports to reduce noise:
```suggestion
                .check_new::<Self>()
                .check_new_service::<Self, http::Request<http::BoxBody>>()
```

cratelyn replied:
db00264b tweaks these to refer to the reexports. thanks for pointing that out, i hadn't realized that `http` in this scope is pointed to `linkerd_app_core::proxy::http`. :bulb:

this looks much better.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route.rs:
```diff
@@ -149,15 +165,6 @@ impl<T: Clone, M, F, P> svc::Param<BackendDistribution<T, F>> for MatchedRoute<T
     }
 }
 
-impl<T: Clone, M, F, P> svc::Param<RouteLabels> for MatchedRoute<T, M, F, P> {
-    fn param(&self) -> RouteLabels {
-        RouteLabels(
-            self.params.parent_ref.clone(),
-            self.params.route_ref.clone(),
-        )
-    }
-}
```
:point_up: here's the core bit driving the deferral of route label extraction, this is no longer some we can know in advance (_because there's a hostname for HTTP metrics_). instead, we use an `ExtractParam<P, T>` to get our labels upon being given a request to inspect.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route/metrics/tests.rs:
```diff
@@ -99,6 +99,127 @@ async fn http_request_statuses() {
     assert_eq!(mixed.get(), 1);
 }
 
+#[tokio::test(flavor = "current_thread", start_paused = true)]
+async fn http_request_hostnames() {
+    const HOST_1: &str = "great.website";
+    const URI_1_1: &str = "https://great.website/path/to/index.html#fragment";
+    const URI_1_2: &str = "https://great.website/another/index.html";
+    const HOST_2: &str = "different.website";
+    const URI_2: &str = "https://different.website/index.html";
+    const URI_3: &str = "https://[3fff::]/index.html";
```
:point_up: tests! we show that our labels look right when sending traffic to different hosts, and to places with an opaque address.

cratelyn commented on linkerd/http/prom/src/record_response/request.rs:
```diff
@@ -72,6 +72,9 @@ where
         (*self.statuses.get_or_create(labels)).clone()
     }
+
+    // TODO(kate): it'd be nice if we could avoid creating a time series if it does not exist,
+    // so that tests can confirm that certain label sets do not exist within the family.
```
i've discussed this in air, but it seems like it'd be a very straightforward change to drive upstream.

cratelyn commented on linkerd/http/retry/src/lib.rs:
```diff
@@ -81,31 +83,45 @@ struct Metrics {
-impl<P, L: Clone, N> NewHttpRetry<P, L, N> {
-    pub fn layer(metrics: MetricFamilies<L>) -> impl layer::Layer<N, Service = Self> + Clone {
+impl<P, X, L: Clone, F: Clone, N> NewHttpRetry<P, L, F, X, N> {
+    pub fn layer_via_mk(
+        mk_extract: F,
+        metrics: MetricFamilies<L>,
+    ) -> impl tower::layer::Layer<N, Service = Self> + Clone {
```
:point_up: bike-shedding welcome. other `layer_via` functions conventionally receive an `X` extractor, in our case we want a function that takes a reference to the `NewService<T>` target (_a matched route_) to construct an `X` that can later extract our `L` labels (_including a hostname_) provided the `Service<T>` target (_a request_).

in a sense this is almost like a `NewExtractParam<..>`, if we can compare it to `NewService`. see the commit message for a little bit more information.

sfleen replied:
It took me a bit to fully grok this (the several layers of tower abstraction are always a bit of a pain to punch through), but I think this is probably the best option given the naming scheme we already have to work with.

cratelyn replied:
https://github.com/linkerd/linkerd2-proxy/pull/3258#discussion_r1794142955 for posterity, some of the changes in this thread helped clarify the layers of :cake: :sparkles: abstraction extraction

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route/retry.rs:
```diff
@@ -29,6 +31,9 @@ pub struct RetryPolicy {
     pub backoff: Option<ExponentialBackoff>,
 }
 
+#[derive(Clone, Debug)]
+pub struct RetryLabelExtract(pub ParentRef, pub RouteRef);
```
our new extractor, which hangs onto the matched route's parent and route metadata.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route/retry.rs:
```diff
@@ -158,3 +163,14 @@ impl RetryPolicy {
+
+// === impl RetryLabelExtract ===
+
+impl<B> ExtractParam<RouteLabels, http::Request<B>> for RetryLabelExtract {
+    fn extract_param(&self, t: &http::Request<B>) -> RouteLabels {
+        let Self(parent, route) = self;
+        let host = t.uri().host();
+
+        RouteLabels::new(parent.clone(), route.clone(), host)
+    }
+}
```
and here, we actually get the hostname, provided the `Service<T>` target: a request with some sort of body.

cratelyn reviewed (commented):
:mag: :bread: breadcrumbs and discussion seeds for reviewers...

sfleen reviewed (approved):
LGTM, I like the overall structure and nothing jumps out at me as too complex given what we're already working with.

cratelyn commented on linkerd/app/outbound/src/http/logical/policy/route.rs:
```diff
@@ -117,20 +116,37 @@ where
                 // TODO(ver) move this into NewSetExtensions?
                 .push(classify::NewClassify::layer())
-                .push(svc::NewMapErr::layer_with(|rt: &Self| {
+                .push(svc::NewMapErr::<_, _, _>::layer_with(|rt: &Self| {
```
ill remove this when im back at a keyboard!

cratelyn replied:
ee5a353d removed this.

olix0r commented on linkerd/http/retry/src/lib.rs:
```diff
@@ -81,31 +83,45 @@ struct Metrics {
-    T: Param<L>,
     P: Policy,
     L: Clone + std::fmt::Debug + Hash + Eq + Send + Sync + prom::encoding::EncodeLabelSet + 'static,
+    F: Fn(&T) -> X + Clone,
```
Eh, this gives me a bit of pause... we're adding an unnamable type parameter here. I think this is basically an ExtractParam?

olix0r replied:
Note https://github.com/linkerd/linkerd2-proxy/blob/89ba5b5d8b39f54de8a8c6e1e8cf552e9c9a5527/linkerd/stack/src/lib.rs#L115-L122

olix0r replied:
To keep things ~consistentish, I'd be inclined to use `X: ExtractParam<T, ReqX>` in place of `F`--since X tends to be used for a when a NewService needs an ExtractParam, and then have a secondary `ReqX` (or something) for the second level of extraction.

cratelyn replied:
750bd6e renames `F` and `X` to `X` and `ReqX`, respectively. 323dedb changes the trait bounds to use the blanket implementation linked above. good call!

olix0r commented on linkerd/http/retry/src/lib.rs:
```diff
@@ -179,21 +195,13 @@ where
     P: Param<Params>,
     P: Clone + Send + Sync + std::fmt::Debug + 'static,
+    L: Clone + std::fmt::Debug + Hash + Eq + Send + Sync + prom::encoding::EncodeLabelSet + 'static,
+    X: ExtractParam<L, http::Request<BoxBody>>,
```
We tend to rely on ExtractParam only for Targets and less so for request parameters... Is https://github.com/linkerd/linkerd2-proxy/pull/3258/commits/69872cb5a5ed06492f023a688bf09978a2ca1411 a change that can be reviewed independently of other changes in this PR?

cratelyn replied:
i've given it a shot, but it seems difficult to extricate that change into its own PR.

to restate my understanding, based on our conversation yesterday, one approach worth trying would be using a request extension rather than an `ExtractParam<P, T>`.

at risk of a brief detour, when we say "Target", we mean the `T` types provided as parameters to e.g. `Http<T>` or `Matched<T>`, at the `NewService<T>` layer. is that right? `Request<BodyBody>` is a concrete target for _this_ `Service<T>`, which had me confused, so i wanted to restate my understanding of this after reading more where we use `ExtractParam`.

i took a swing at this, here: https://github.com/linkerd/linkerd2-proxy/tree/kate/add-hostname-label-to-http-metrics.failed-extension-attempt

i found myself running into a lot of trouble with this, unfortunately. there's still something slightly incongruent stemming from the fact that we have labels that straddle the boundary between a construction time (_route and parent references_) and call-time (_request `Parts`_).

cratelyn replied:
we resolved this in-air, and confirmed the "target" terminology. :heavy_check_mark:

cratelyn commented:
pushed some commits addressing some of the low-hanging fruit, there is still some feedback to address... :apple: :white_check_mark:

olix0r commented on linkerd/app/outbound/src/http/logical/policy/route.rs:
```diff
@@ -117,16 +116,33 @@ where
                 .push(filters::NewApplyFilters::<Self, _, _>::layer())
-                .push(retry::NewHttpRetry::layer(metrics.retry.clone()))
+                .push({
+                    let mk_extract = |rt: &Self| {
```
Let's add a TODO here to note that this extractions of labels probably belongs somewhere else (like SetExtensions).

olix0r commented on linkerd/app/outbound/src/http/logical/policy/route/metrics/labels.rs:
```diff
@@ -1,21 +1,29 @@
 
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
-pub struct Route(pub ParentRef, pub RouteRef);
+pub struct HttpRoute {
+    parent: ParentRef,
+    route: RouteRef,
+    name: Option<dns::Name>,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct GrpcRoute(pub ParentRef, pub RouteRef);
```
I think we should leave these as unified so that gRPC routes also include a hostname.

olix0r commented on linkerd/app/outbound/src/http/logical/policy/route/metrics/labels.rs:
```diff
@@ -1,21 +1,29 @@
+pub struct HttpRoute {
+    parent: ParentRef,
+    route: RouteRef,
+    name: Option<dns::Name>,
```
'name' can be a bit ambiguous in the context of our code... especially with regard to identity vs hosts vs ...

```suggestion
    hostname: Option<dns::Name>,
```

olix0r commented on linkerd/http/retry/src/lib.rs:
```diff
@@ -223,6 +231,8 @@ where
         };
 
         let params = policy.param();
+        let labels = self.extract.extract_param(&req);
```
Perhaps worth a TODO that we'd like to avoid knowing about labels here in the future.

cratelyn replied:
added in 1c03f91 :hammer:

cratelyn commented:
**note:** i have squashed this branch into a single commit, due to the fact that some changes were walked back in the process of review. https://github.com/linkerd/linkerd2-proxy/tree/kate/add-hostname-label-to-http-metrics.pre-squash-24-10-14 this branch has been pushed for posterity, which contains these changes and incremental review commits left intact.

after squashing, i've also rebased this branch on top of the latest state of `main`.

cratelyn commented:
:curly_loop: tying up a final loop, updated the PR description above, and the commit message to be up-to-date with what we're landing now that we've gone through review. this should now be ready to ship :shipit:

olix0r reviewed (approved):
🌮
